ow_lander/scripts/trajectory_publisher.py

```diff
@@ -18,27 +19,39 @@
 import numpy as np
 from ow_lander.msg import GuardedMoveResult
 
-from peak_detection_real_time import PeakDetectionRT
+# TODO: rather than basing the decision on a single value we can take the trend
+# over a short term of position z value it would probably make the implementation
+# more robust against an arm physics configuration where the end effector doesn't
+#  bounce off the ground at all.
 
-lag = 80
-threshold = 6.3
-influence = 0.5
-peak_detection = PeakDetectionRT(lag=lag, threshold=threshold, influence=influence)
+# constanst
+GROUND_DETECTION_THRESHOLD = 0.001  # TODO: tune this value further to improve the detection
```

Comment:
This should probably factor in timestep in case we want to run the sim at different speeds in the future. For example GROUND_DETECTION_THRESHOLD = 0.5 * timestep

This effectively makes the detection based on velocity. Someone correct me if I'm wrong: I thought we switched away from velocity because it's an indirect measurement we can only estimate in real-life using joint states. That said, I'm fine with "cheating" and using ground truth velocity if its the only way forward in the short term.

Reply:
This solution is already effectively based on velocity since I am basing my decision on the delta of two positions.

With regard to multiplying by a timestep that should be fine however I think the proper way is to have monitor the trend of velocities.

_Note_: there is no cheating in this solution, I can use tf2 to read the change in position over the cycle and the solution wouldn't change much, but gazebo/link_states gives already provides me with a stream of joints positions it would much more efficient implementation this way.

Again, no cheating in this approach.

Reply:
It is more efficient than using tf2 because I don't have to pull the joint values.


The only place this would be cheating is if we want the solution to work outside of the simulation i.e on hardware robot.

Reply:
This solution is based on distance, not velocity. Changing the timestep of the sim could affect this solution. I hope I didn't offend too much with the word cheating. I only mean to say it's using ground truth info that a real robot wouldn't have.

Reply:
I did address the cheating topic (not responding because of feeling offended). Before implementing I thought of using **tf2** but resorted to using **gazebo/link_states** mainly favoring to subscribe to the available stream of values rather than having to pull these values out of tf2.

   

> This solution is based on distance, not velocity.
the computed delta is between two readings taken at two different times, the system clock is not factored but that's a different thing.

Reply:
In any case I will provide an implementation based off tf2 (while keeping the gazebo/link_states until I fully understand the motives)

Reply:
Actually, I'm assuming this script runs once per gazebo timestep. If it's running on a separate clock then it might not matter.

Reply:
I did not say that factoring time delta wouldn't be useful. I was just trying to explain that what I have here is technically velocity. I haven't put enough thought into the GROUND_DETECTION_THRESHOLD value.

Reply:
addressed